File: cpp/jamfree/python/examples/highway_simulation.py
#!/usr/bin/env python3
"""
Highway simulation example using JamFree Python bindings.

This example demonstrates:
- Creating a multi-lane highway
- Adding vehicles with different speeds
- Using IDM car-following model
- Using MOBIL lane-changing model
- Running a traffic simulation
"""

import jamfree
import time

def main():
    print("=" * 60)
    print("JamFree Python - Highway Simulation with Lane Changing")
    print("=" * 60)
    print()

    # Create a 3-lane highway
    highway = jamfree.Road(
        "highway_1",
        jamfree.Point2D(0, 0),
        jamfree.Point2D(3000, 0),  # 3 km
        3,    # 3 lanes
        3.5   # 3.5m lane width
    )
    
    print(f"Created: {highway}")
    print()

    # Create IDM car-following model
    idm = jamfree.IDM(
        desired_speed=jamfree.kmh_to_ms(120),  # 120 km/h
        time_headway=1.5,
        min_gap=2.0,
        max_accel=1.0,
        comfortable_decel=1.5
    )
    
    print(f"Car-following model: {idm}")
    
    # Create MOBIL lane-changing model
    mobil = jamfree.MOBIL(
        politeness=0.3,
        threshold=0.1,
        max_safe_decel=4.0,
        bias_right=0.2
    )
    
    print(f"Lane-changing model: {mobil}")
    print()

    # Create vehicles
    vehicles = []
    
    # Slow truck in right lane
    truck = jamfree.Vehicle("truck", length=12.0)
    truck.set_current_lane(highway.get_lane(0))
    truck.set_lane_position(500.0)
    truck.set_speed(jamfree.kmh_to_ms(80))
    highway.get_lane(0).get_vehicles()  # This adds it to the lane
    vehicles.append(truck)
    print(f"Added: {truck} in lane 0")
    
    # Fast cars behind truck
    for i in range(5):
        car = jamfree.Vehicle(f"car_{i}")
        car.set_current_lane(highway.get_lane(0))
        car.set_lane_position(100.0 + i * 50.0)
        car.set_speed(jamfree.kmh_to_ms(100 + i * 2))
        vehicles.append(car)
        print(f"Added: {car} in lane 0")
    
    print()
    print("Running simulation for 60 seconds...")
    print()

    # Simulation parameters
    dt = 0.1  # Time step: 0.1s
    num_steps = 600  # 60 seconds
    report_interval = 100  # Report every 10 seconds

    # Simulation loop
    for step in range(num_steps):
        # Phase 1: Update vehicle dynamics (IDM)
        for vehicle in vehicles:
            lane = vehicle.get_current_lane()
            if not lane:
                continue
            
            # Find leader (simplified - would need proper implementation)
            leader = None  # In real implementation, find leader in lane
            
            # Calculate acceleration
            accel = idm.calculate_acceleration(vehicle, leader)
            
            # Update vehicle
            vehicle.update(dt, accel)
        
        # Phase 2: Lane changing (MOBIL)
        for vehicle in vehicles:
            current_lane = vehicle.get_current_lane()
            if not current_lane:
                continue
            
            lane_idx = current_lane.get_index()
            
            # Get adjacent lanes
            left_lane = highway.get_lane(lane_idx + 1) if lane_idx < 2 else None
            right_lane = highway.get_lane(lane_idx - 1) if lane_idx > 0 else None
            
            # Decide lane change
            decision = mobil.decide_lane_change(
                vehicle, current_lane, 
                left_lane, right_lane, 
                idm
            )
            
            # Execute lane change
            if decision == jamfree.LaneChangeDirection.LEFT and left_lane:
                print(f"  [t={step*dt:.1f}s] {vehicle.get_id()} changes LEFT "
                      f"(lane {lane_idx} -> {lane_idx+1})")
                vehicle.set_current_lane(left_lane)
                
            elif decision == jamfree.LaneChangeDirection.RIGHT and right_lane:
                print(f"  [t={step*dt:.1f}s] {vehicle.get_id()} changes RIGHT "
                      f"(lane {lane_idx} -> {lane_idx-1})")
                vehicle.set_current_lane(right_lane)
        
        # Periodic status report
        if step % report_interval == 0 and step > 0:
            print()
            print(f"Status at t={step*dt:.1f}s:")
            for vehicle in vehicles:
                lane = vehicle.get_current_lane()
                lane_idx = lane.get_index() if lane else -1
                print(f"  {vehicle.get_id():12s}: lane={lane_idx}, "
                      f"pos={vehicle.get_lane_position():7.1f}m, "
                      f"speed={jamfree.ms_to_kmh(vehicle.get_speed()):5.1f} km/h")
            print()

    # Final state
    print()
    print("Final State (t=60s):")
    for vehicle in vehicles:
        lane = vehicle.get_current_lane()
        lane_idx = lane.get_index() if lane else -1
        print(f"{vehicle.get_id()}:")
        print(f"  Lane: {lane_idx}")
        print(f"  Position: {vehicle.get_lane_position():.1f} m")
        print(f"  Speed: {jamfree.ms_to_kmh(vehicle.get_speed()):.1f} km/h")
        print(f"  Acceleration: {vehicle.get_acceleration():.2f} m/s²")
        print()

    print("=" * 60)
    print("Simulation complete!")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: cpp/jamfree/examples/highway_simulation.cpp
#include "../microscopic/include/agents/VehiclePrivateLocalStateMicro.h"
#include "../microscopic/include/agents/VehiclePublicLocalStateMicro.h"
#include "../microscopic/include/decision/VehicleDecisionModelMicro.h"
#include "../microscopic/include/IDM.h"
#include "../microscopic/include/decision/dms/SubsumptionDMS.h"
#include "../microscopic/include/decision/dms/ForwardAccelerationDMS.h"
#include "../microscopic/include/perception/VehiclePerceptionModelMicro.h"
#include "../microscopic/include/reaction/MicroscopicReactionModel.h"
#include "../kernel/include/agents/VehicleAgent.h"
#include "../kernel/include/model/Lane.h"
#include "../kernel/include/model/Road.h"
#include "../kernel/include/simulation/SimulationEngine.h"
#include <iostream>
#include <memory>

using namespace jamfree;

/**
 * @brief Simple highway traffic simulation example.
 *
 * This example demonstrates:
 * - Creating a highway with multiple lanes
 * - Adding vehicles with different behaviors
 * - Running a simulation with perception, decision, and reaction
 * - Visualizing the results
 */
int main() {
  std::cout << "=== JamFree Highway Simulation Example ===" << std::endl;
  std::cout << "Using SIMILAR Extended Kernel Architecture" << std::endl
            << std::endl;

  // Create simulation engine
  double timeStep = 0.1; // 100ms time step
  auto engine = std::make_shared<kernel::simulation::SimulationEngine>(timeStep);

  // Create a 3-lane highway (1km long)
  kernel::model::Point2D start(0, 0);
  kernel::model::Point2D end(1000, 0);
  auto highway = std::make_shared<kernel::model::Road>("Highway-1", start, end, 3, 3.5);

  std::cout << "Created highway: " << highway->getId() << std::endl;
  std::cout << "  Lanes: " << highway->getNumLanes() << std::endl;
  std::cout << "  Length: " << start.distanceTo(end) << " meters" << std::endl
            << std::endl;

  // Create reaction model for microscopic level
  auto reactionModel =
      std::make_shared<microscopic::reaction::MicroscopicReactionModel>(timeStep);
  kernel::agents::LevelIdentifier microLevel("Microscopic");
  engine->setReactionModel(microLevel, reactionModel);

  // Create vehicles
  std::cout << "Creating vehicles..." << std::endl;

  for (int i = 0; i < 5; ++i) {
    // Create vehicle agent
    std::string vehicleId = "Vehicle-" + std::to_string(i);
    auto vehicle = std::make_shared<kernel::agents::VehicleAgent>(vehicleId);

    // Create public state (visible to all)
    auto publicState =
        std::make_shared<microscopic::agents::VehiclePublicLocalStateMicro>(vehicleId);
    publicState->setCurrentLane(highway->getLane(i % 3).get());
    publicState->setLaneIndex(i % 3);
    publicState->setLanePosition(i * 50.0);  // Stagger vehicles
    publicState->setSpeed(25.0 + (i * 2.0)); // Different speeds (90-108 km/h)
    publicState->setAcceleration(0.0);
    publicState->setActive(true);

    // Create private state (agent-specific)
    auto privateState =
        std::make_shared<microscopic::agents::VehiclePrivateLocalStateMicro>(vehicleId);
    privateState->setDesiredSpeed(30.0 + (i * 2.0)); // Desired speed
    privateState->setTimeHeadway(1.5);
    privateState->setMinGap(2.0);
    privateState->setMaxAcceleration(2.0);
    privateState->setComfortableDeceleration(3.0);
    privateState->setPoliteness(0.5);

    // Create perception model
    auto perceptionModel =
        std::make_shared<microscopic::perception::VehiclePerceptionModelMicro>(150.0);

    // Create decision model with subsumption DMS
    auto subsumptionDMS = std::make_shared<microscopic::decision::dms::SubsumptionDMS>();

    // Add IDM-based acceleration DMS
    auto idm = std::make_shared<microscopic::models::IDM>();
    auto accelDMS = std::make_shared<microscopic::decision::dms::ForwardAccelerationDMS>(idm);
    subsumptionDMS->addSubmodel(accelDMS);

    auto decisionModel =
        std::make_shared<microscopic::decision::VehicleDecisionModelMicro>(subsumptionDMS);

    // Add to agent
    vehicle->includeNewLevel(microLevel, publicState, privateState);
    vehicle->setPerceptionModel(microLevel, perceptionModel);
    vehicle->setDecisionModel(microLevel, decisionModel);

    // Add to simulation
    engine->addAgent(vehicle);

    std::cout << "  " << vehicleId << ": Lane " << (i % 3) << ", Position "
              << (i * 50.0) << "m, Speed " << publicState->getSpeed() << " m/s"
              << std::endl;
  }

  std::cout << std::endl << "Running simulation..." << std::endl;
  std::cout << "Time Step: " << timeStep << "s" << std::endl;
  std::cout << "Duration: 10 seconds (100 steps)" << std::endl << std::endl;

  // Run simulation for 10 seconds
  int numSteps = 100;
  for (int step = 0; step < numSteps; ++step) {
    engine->step();

    // Print status every 10 steps (1 second)
    if (step % 10 == 0) {
      std::cout << "t=" << engine->getCurrentTime() << "s: ";

      const auto &agents = engine->getAgents();
      for (const auto &agent : agents) {
        auto publicState = agent->getPublicLocalState(microLevel);
        if (publicState) {
          auto vehicleState =
              std::dynamic_pointer_cast<microscopic::agents::VehiclePublicLocalStateMicro>(
                  publicState);
          if (vehicleState) {
            std::cout << agent->getId() << "="
                      << static_cast<int>(vehicleState->getLanePosition())
                      << "m ";
          }
        }
      }
      std::cout << std::endl;
    }
  }

  std::cout << std::endl << "Simulation complete!" << std::endl;
  std::cout << "Total steps: " << engine->getStepCount() << std::endl;
  std::cout << "Final time: " << engine->getCurrentTime() << "s" << std::endl;

  return 0;
}
